### PR TITLE
[MM-55485] Ensure user added to channel receives websocket event regardless of channel membership info retrieved in HA

### PR DESCRIPTION
**Pull request by devinbinnie**
#### Summary
There have been numerous reports of users missing messages from channels they've been added to. These reports were all from our Cloud customers or other users in HA, community included. What was found was that the `user_added` event was to be sent to members of the channel, which in theory should include the newest added user when the websocket event was scoped to the channel. However, in an HA environment, there's no specific guarantee that the channel members were updated yet, and when the websocket module goes to fetch them, they get the older copy (ie. a read-after-write issue)

This might speak to some work that can be done in the cluster syncing, but for this particular issue, we can simply just scope the websocket event to the specific user that was added, ensuring that regardless of perceived channel membership (and we do know that they are a channel member now) they should receive the websocket event.

This PR splits the call into two portions: one event sent to the rest of the channel who are already members, omitted the newly added user, and one sent specifically to the new user. This should alleviate any HA issue that could be causing this. We will want to test in community to see if it is still reproducible at all.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-55485

```release-note
NONE
```

### CODE REVIEW

**Review by agnivade — COMMENTED** [review]
So this is what happens when you add a user to the channel:
1. `a.addUserToChannel` is called.
2. That does its thing and eventually calls `a.InvalidateCacheForUser(user.Id)`. This zaps the webConn channel member cache which it uses to check if a user is part of a channel.
3. Now it tries to publish the event. Inside the webconn code, it notices that the channel member cache is gone, so it tries to fetch it from the DB.
```go
if wc.allChannelMembers == nil {
			result, err := wc.Platform.Store.Channel().GetAllChannelMembersForUser(wc.UserId, false, false)
	...
```
4. Now here, it queries the replica and can miss the new member because of replica lag.

This is a common problem that we have run into in the past, and the recommended solution is to pass a context to the store method to always query master to refresh the channel member cache.

Check `(s SqlTeamStore) GetTeamsForUser` for a sample implementation. You need to pass `RequestContextWithMaster(c)` to the store method.

This will avoid weirdly structuring code like this and keeps things simple.

**Review by agnivade — COMMENTED** [review]
Just a small comment: whenever we are declaring a new variable `ctx`, let's rename it to `rctx`. This is a convention we agreed on to differentiate between context.Context and request.CTX. There will need to be a larger effort to rename all such variables to `rctx` but we will do it at a later point.

**Comment by devinbinnie** [comment]
@agnivade I'm not sure the solution above is working - I'm still very much able to reproduce this issue with the current approach of syncing to the DB master. Do I need to do it in other places as well?

**Comment by devinbinnie** [comment]
@agnivade I can reproduce with master hardcoded. Would it make sense to use my original solution for now since it seems to circumvent any cluster syncing issues?

I'm not really sure where to start otherwise.

**Comment by agnivade** [comment]
Sure, although it would be good to understand what the problem is really. As it stands, we aren't sure where the problem exactly is, which is not a good thing to have.

File: server/channels/app/channel.go [hunk 1]
```diff
@@ -1597,11 +1597,16 @@ func (a *App) AddUserToChannel(c request.CTX, user *model.User, channel *model.C
 		return nil, err
 	}
 
-	message := model.NewWebSocketEvent(model.WebsocketEventUserAdded, "", channel.Id, "", nil, "")
+	message := model.NewWebSocketEvent(model.WebsocketEventUserAdded, "", channel.Id, "", map[string]bool{user.Id: true}, "")
```
Comment (agnivade) [inline server/channels/app/channel.go:1600]: Please leave a comment explaining why we are doing it this way.

**Review by agnivade — APPROVED** [review]
Please leave

**Review by hmhealey — APPROVED** [review]

**Pull request merged**